src/CTA.Rules.RuleFiles/RulesFileParser.cs

```diff
@@ -156,9 +156,18 @@ private void ProcessObject(Rootobject rootobject)
                     {
                         if (method.Actions != null && method.Actions.Count > 0)
                         {
-                            var token = new InvocationExpressionToken() { Key = method.Key, Namespace = @namespace.@namespace, FullKey = method.FullKey, Type = @class.Key };
-                            if (!_rootNodes.Invocationexpressiontokens.Contains(token)) { _rootNodes.Invocationexpressiontokens.Add(token); }
-                            ParseActions(token, method.Actions);
+                            if (@class.KeyType == CTA.Rules.Config.Constants.InvocationExpressionActions)
+                            {
+                                var token = new InvocationExpressionToken() { Key = method.Key, Namespace = @namespace.@namespace, FullKey = method.FullKey, Type = @class.Key };
+                                if (!_rootNodes.Invocationexpressiontokens.Contains(token)) { _rootNodes.Invocationexpressiontokens.Add(token); }
+                                ParseActions(token, method.Actions);
+                            }
+                            if (@class.KeyType == CTA.Rules.Config.Constants.ExpressionActions)
+                            {
+                                var token = new ExpressionToken() { Key = method.Key, Namespace = @namespace.@namespace, FullKey = method.FullKey, Type = @class.Key };
+                                if (!_rootNodes.Expressiontokens.Contains(token)) { _rootNodes.Expressiontokens.Add(token); }
+                                ParseActions(token, method.Actions);
+                            }
```

Comment:
Why are we setting this inside the invocation expression instead of creating a different type?

Reply:
The only type of expressions that would receive the await keyword would be of type invocation expressions. So it felt natural to put it there. Furthermore any of the other possible node exploration methods did not fit the type of node that are needed to add await to (namely invocation expression types of nodes).

Reply:
After reviewing together I understand the goal to move file parser logic to flat structure rather than hierarchical one. Also implemented a new Visit node function for ExpressionStatementSyntax nodes.